src/lexparse/lexer.go

```diff
@@ -3,73 +3,228 @@ package lexparse
+
+  for i, s := range lines_raw {
+    lines[i] = Line{
+      Num: i,
+      Value: []rune(s),
+      Filename: filename}
+  }
+
+  tokens := []Token{}
+  lexer := Lexer{InBlockComment: false}
+
+  for _, line := range lines {
+    lexer.startline(&line)
+    tokens = append(tokens, lexer.grabTokens()...)
+  }
+
+  return tokens
+}
+
+func (lexer *Lexer) startline(line *Line) {
+  lexer.Line = line
+  lexer.Col = 0
+}
+
+func (lexer *Lexer) newToken() {
+  lexer.CurrentBlob = []rune{}
+  lexer.CurrentTType = -1
+}
+
+func (lexer *Lexer) next() rune {
+  if !lexer.hasMore() { return rune(0) }
+  return lexer.Line.Value[lexer.Col]
+}
+
+func (lexer *Lexer) advance() {
+  lexer.CurrentBlob = append(lexer.CurrentBlob, lexer.next())
+  lexer.Col++
+}
+
+func (lexer *Lexer) hasMore() bool {
+  return lexer.Col + 1 < len(lexer.Line.Value)
 }
 
-func GrabTokens(line Line) []Token {
-  t := Token{line, 4, SymbolT, line.Value[4:5]}
-  lexerPanic(t, "Oh noes! ur a bad programmer! I'm gonna tell ur mom!")
-  return []Token{t}
+func (lexer *Lexer) passWhitespace() {
+  for unicode.IsSpace(lexer.next()) {
+    lexer.advance()
+  }
 }
 
-// func GrabToken(line Line, col uint8) {
-// }
+func (lexer *Lexer) currentToken() Token {
+  return Token {
+    Line: lexer.Line,
+    Col: lexer.Col - len(lexer.CurrentBlob),
+    TType: lexer.CurrentTType,
+    Value: lexer.CurrentBlob,
+  }
+}
+
+func (lexer *Lexer) grabTokens() []Token {
+  tokens := []Token{}
+  lexer.passWhitespace()
+
+  for lexer.hasMore() {
+    tokens = append(tokens, lexer.grabToken())
+    lexer.passWhitespace()
+  }
+
+  return tokens
+}
+
+func (lexer *Lexer) grabToken() Token {
+  lexer.newToken()
+
+  c := lexer.next()
+  if unicode.IsLetter(c) || (c == '_') {
+    lexer.grabSymbol()
+  } else if unicode.IsDigit(c) {
+    lexer.grabDecimalNumber()
+  } else {
+    lexer.grabPunctuation()
+  }
+
+  return lexer.currentToken()
+}
+
+func (lexer *Lexer) grabSymbol() {
+  c := lexer.next()
+  for unicode.IsLetter(c) || unicode.IsDigit(c) || (c == '_') {
+    lexer.advance()
+    c = lexer.next()
+  }
+
+  switch symbol := string(lexer.CurrentBlob); symbol {
+    case "true":  lexer.CurrentTType = BoolT
+    case "false": lexer.CurrentTType = BoolT
+    case "null":  lexer.CurrentTType = NullT
+    case "for":   lexer.CurrentTType = ForT
+    case "while": lexer.CurrentTType = WhileT
+    case "in":    lexer.CurrentTType = InT
+    case "type":  lexer.CurrentTType = TypeT
+    default:      lexer.CurrentTType = SymbolT
+  }
+}
+
+func (lexer *Lexer) grabDecimalNumber() {
+  for unicode.IsDigit(lexer.next()) {
+    lexer.advance()
+  }
+
+  if lexer.next() != '.' {
+    lexer.CurrentTType = IntT
+  } else {
+    lexer.CurrentTType = FloatT
+
+    lexer.advance()
+
+    for unicode.IsDigit(lexer.next()) {
+      lexer.advance()
+    }
+  }
+}
+
+func (lexer *Lexer) grabPunctuation() {
+  lexer.advance()
+
+  var blob string
+  twochars := string(append(lexer.CurrentBlob, lexer.next()))
```

Comment:
Does this cause any problems if the character immediately following punctuation is not another punctuator? E.g.

```
if x <y 
```

I ask because it looks like we're advancing the lexer here without checking if the advance is valid.

Reply:
Ah, `next()` doesn't actually advance the lexer, so this is just a lookahead. However, you've alerted me to the fact that an `advance()` needs to make it into the `blob = twochars` codeblock!